lib/WeBWorK3/Controller/Permission.pm

```diff
@@ -0,0 +1,99 @@
+package WeBWorK3::Controller::Permission;
+use warnings;
+use strict;
+
+use Mojo::Base 'Mojolicious::Controller', -signatures;
+
+sub getRoles ($c) {
+	my @roles_as_rs = $c->schema->resultset('Role')->search;
+	# convert to hashes
+	my @roles = map { $_->role_name } @roles_as_rs;
+	$c->render(json => \@roles);
+	return;
+}
+
+sub getUIRoutePermissions ($c) {
+	my @perms       = $c->schema->resultset('UIPermission')->search;
+	my @route_perms = map {
+		{ $_->get_inflated_columns };
+	} @perms;
+	$c->render(json => \@route_perms);
+	return;
+}
+
+# The checkPermission subroutine checks if the passed in user has permission
+# to access the route given in the controller/action.
+#
+# This is done first by
+# 1. ensuring that a route exists in the database.
+# 2. checking if the user is an admin and if so, allowing them in.
+# 3. if the route just needs authentication access, permit the user.
+# 4. return false if the route requires admin access and the user is not an admin.
+# 5. if the route allows self access return true if the user_id matches the route.
+# 6. checking that the user has a role with an allowed permission.
+#
+
+sub checkPermission ($c) {
+	# seems like $c->current_user doesn't have access to the course_id, so we never
+	# get a CourseUser.
+	my $user = $c->current_user;
+
+	# The controller/action pair is stored in the $c->match->stack array at the end.
+	# See https://docs.mojolicious.org/Mojolicious/Guides/Routing#Under
+	my $controller = $c->match->stack->[1]->{controller};
+	my $action     = $c->match->stack->[1]->{action};
```

Comment:
There is unnecessary de-referencing here.  This should be
```suggestion
	my $controller = $c->match->stack->[1]{controller};
	my $action     = $c->match->stack->[1]{action};
```
Even the Mojolicious documentation you referenced does it this way.

There is more of this below in this file (and probably elsewhere).

@dpvc pointed this sort of thing out to me a few years ago.  webwork2 has massive amounts of this sort of unnecessary de-referencing.

Reply:
@pstaabp and I talked about this -- yes, we have absolutely picked up some bad habits from the ww2 codebase.